proposals/NNNN-multiple-trailing-closures.md

```diff
@@ -0,0 +1,378 @@
+syntax to cover all of the use cases that are awkward in the language today. A set of curly braces delimits
+the block of trailing closures, with each trailing closure indicated by an argument label inside.
+For example, Ehab's example becomes:
+
+```swift
+UIView.animate(withDuration: 0.7, delay: 1.0, options: .curveEaseOut) {
+  animations: {
+    self.view.layoutIfNeeded()
+  }
+  completion: { finished in
+    print("Basket doors opened!")
+  }
+}
+```
+
+Similar to computed properties, the outer curly braces syntactically hold all of the trailing closures
+together. Within those curly braces are argument label-closure pairs to specify the trailing closures,
+which makes it clear what the purpose of each closure is. The call site also nicely separates
+the "configuration" arguments (how to animate) from the "action" arguments (what to do at each step).
+This syntax handles nesting cleanly; consider Anupam's example again, now with multiple trailing
+closures:
+
+```swift
+func toggle() {
+  UIView.animate(withDuration: 1) {
+    animations: {
+      self.myView.backgroundColor = UIColor.green
+      self.myView.frame.size.width += 50
+      self.myView.frame.size.height += 20
+      self.myView.center.x += 20
+    }
+    completion: { _ in
+      UIView.animate(withDuration: 1, delay: 0.25, options: [.autoreverse, .repeat]) {
+        animations: {
+          self.myView.frame.origin.y -= 20
+         }
+       }
+     }
+   }
+}
+```
+
+Here, we've eliminated the inconsistencies where some "calls" close with a single `}` and others
+with a `})`. The new syntax composes well with SwiftUI's `Button`:
+
+```swift
+Button {
+  action: {
+    ...
+    ...
+    ...
+  }
+
+  label: {
+    Text("Hello!")
+  }
+}
+```
+
+and our `when(then:else:)` control-flow construct:
+
+```swift
+when(2 < 3) {
+   then: {
+     ...
+     ...
+   }
+   else: {
+     ...
+     ...
+   }
+}
+```
+
+Note that none of the APIs used here need to change. This proposal improves
+[clarity at the point of use](https://swift.org/documentation/api-design-guidelines/#fundamentals) for
+many existing APIs, without requiring the API authors to make changes.
+
+## Detailed design
+
+Changes required to implement new syntax are constrained solely to the parser, namely to the parsing of call arguments.
+
+It's possible to do an early syntax transformation which would consider each block to be a regular argument and adjust locations of parentheses if necessary.
+For type checker perspective this means that calls with new syntax would just be regular calls which require no special handling:
+
+```swift
+when(2 < 3) {
+   then: { ... }
+   else: { ... }
+}
+```
+
+Would be transformed into:
+
+```swift
+when(2 < 3,
+    then: { ... },
+    else: { ... }
+)
+```
+
+Since all of the essential source information is preserved (locations of all labels and closure blocks) it would be possible for type-checker to produce diagnostics for invalid code without any changes.
```

Comment:
> *the* type-checker

```suggestion
Since all of the essential source information is preserved (locations of all labels and closure blocks) it would be possible for the type-checker to produce diagnostics for invalid code without any changes.
```